src/include/simeng/PerceptronPredictor.hh

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <deque>
+#include <map>
+#include <vector>
+
+#include "simeng/BranchPredictor.hh"
+#include "simeng/config/SimInfo.hh"
+
+namespace simeng {
+
+/** A Perceptron branch predictor implementing the branch predictor described in
+ * Jimenez and Lin ("Dynamic branch prediction with perceptrons", IEEE High-
+ * Performance Computer Architecture Symposium Proceedings (2001), 197-206.
+ * The following predictors have been included:
+ *
+ * - Static predictor based on pre-allocated branch type.
+ *
+ * - A Branch Target Buffer (BTB) with a local and global indexing scheme and a
+ * perceptron.
+ *
+ * - A Return Address Stack (RAS) is also in use.
+ */
+
+class PerceptronPredictor : public BranchPredictor {
+ public:
+  /** Initialise predictor models. */
+  PerceptronPredictor(ryml::ConstNodeRef config = config::SimInfo::getConfig());
+  ~PerceptronPredictor();
+
+  /** Generate a branch prediction for the supplied instruction address, a
+   * branch type, and a known branch offset; defaults to 0 meaning offset is not
+   * known. Returns a branch direction and branch target address. */
+  BranchPrediction predict(uint64_t address, BranchType type,
+                           int64_t knownOffset = 0) override;
+
+  /** Updates appropriate predictor model objects based on the address and
+   * outcome of the branch instruction. */
+  void update(uint64_t address, bool taken, uint64_t targetAddress,
+              BranchType type) override;
+
+  /** Provides RAS rewinding behaviour. */
+  void flush(uint64_t address) override;
+
+ private:
+  /** The bitlength of the BTB index; BTB will have 2^bits entries. */
+  uint64_t btbBits_;
+
+  /** A 2^bits length vector of pairs containing a (globalHistoryLength_ + 1)
+   * length perceptron and a branch target. */
+  std::vector<std::pair<std::vector<int8_t>, uint64_t>> btb_;
+
```

Comment:
Is branch target buffer the best name for this structure? "Table of perceptrons" therefor "top_" from the paper may be more descriptive. 

This also relates to the config options - other than the "Type" option, it isn't obvious how the other options are effecting anything to do with perceptrons e.g. "BTB-Tag-Bits" is setting the size of the "top_" and therefor the number of perceptrons in use. In the case of DEMO_RISCV that is 2^11=2048 perceptrons which isn't obvious when just looking at the config.

Reply:
Maybe this should be separated out into a top_ and a btb_. btb_[n].second more aligns with the idea of a branch target buffer

Reply:
This language is carried over from the GenericPredictor, which also uses "btb_" to describe a table containing the targets and the prediction mechanism (in that case a saturating counter).  I think separating it out into two different tables feels clunky as they would both be indexed with the same hash and the pair of data currently in each entry is symmetrical to the pair of data provided in the prediction object that is returned by predict() (i.e. both contain direction information and target information).  However, very happy to switch up the naming -- maybe something like predictorTable_?

Reply:
The reason _for_ separating into 2 different structures is that it would more accurately represent the hardware as a BTB is simply a cache for the target addresses and nothing else. This would also align the implementation more to the literature making it easier to understand. 

The reason _against_ separating is a simpler, "less clunky" and potentially slightly quicker SimEng implementation (although would need to verify that) so I can absolutely see why it is done that way. 

We don't necessarily need to align with hardware at this level so I am fine with the way it is implemented (e.g. 1 vector containing prediction mechanism and target), however I think the name is confusing when the structure is doing more than what a BTB (in the literature) is supposed to do. In that regard I am also not fully comfortable with the name in the generic predictor either. Interested in @jj16791 's opinion as he did this implementation.

"predictorTable_" could be a better name which could also be used in the generic predictor.

Reply:
Having read the docs Alex wrote, I think BTB is fine. I get your point on this @dANW34V3R, but given our usage of the BTB still "caches" targets (plus some more information) it lives up to its name.

I'm not fussed either way if its just a name change (I don't feel as though its required, but I'm not against it), but I don't agree with splitting this into 2 structures

Reply:
So my opinions on both matters raised are the following.

I agree with Alex regarding not splitting into two different tables and the supporting point Dan made about not having to align with hardware is also spot on regarding SimEng's overall view of things like this.

As for the naming, both structures in our two Branch Predictors do represent Branch Target Buffers but with different predicting logic. BTB's are a very common and accessible structure regarding the understanding of its functionality so it's a good naming convention to use here. The generic predictor's saturating counter will also be well-understood so I don't think much needs to change there. The perception usage here is a bit less familiar at large in the domain we're in so perhaps some more verbose commenting above btb_ would be beneficial. We do link in the resources to fully understand the Branch Predictor at a detailed level so any additional information should just give a high-level overview of terminology/algorithms in use rather than in-depth detailings.

Reply:
I'm still in favour of a name change e.g. predictorTable_ as Alex suggests as getting the prediction mechanism from a buffer containing branch targets doesn't make sense. Can discuss fully in the meeting also to save time here?

Reply:
Offline discussion agreement to keep the name but use paper accurate language to the comments and docs. Will resolve once complete